## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [4]:
import pandas as pd

df = pd.read_csv("/Users/michaelbrady/Desktop/working/pandas-challenge-1/Resources/client_dataset.csv")
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [5]:
# View the column names in the data
df.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [6]:
# Use the describe function (accidentally deleted this cell..)
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [7]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.shape

(54639, 18)

In [8]:
df.isnull().sum()

first          0
last           0
job            0
phone          0
email          0
client_id      0
order_id       0
order_date     0
order_week     0
order_year     0
item_id        0
category       0
subcategory    0
unit_price     0
unit_cost      0
unit_weight    0
qty            0
line_number    0
dtype: int64

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [10]:
# What three item categories had the most entries?
top_3_cat = df["category"].value_counts().head(3)
top_3_cat

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [11]:
# For the category with the most entries, which subcategory had the most entries?
top_subcat = df[df["category"] == "consumables"]["subcategory"].value_counts().idxmax()
top_subcat

'bathroom supplies'

In [12]:
# Which five clients had the most entries in the data?
top_5_clients = df["client_id"].value_counts().head(5)
top_5_clients

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [13]:
# Store the client ids of those top 5 clients in a list.
top_5_list = top_5_clients.index.tolist()
top_5_list

[33615, 66037, 46820, 38378, 24741]

In [14]:
# How many total units (the qty column) did the client with the most entries order?
top_client = top_5_clients.index[0]
top_total_units = df[df["client_id"] == top_client]["qty"].sum()
top_total_units

np.int64(64313)

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [15]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df["line_subtotal"] = df["unit_price"] * df["qty"]
df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year         item_id     category  \
0              17        2023  EUD29711-63-6U        decor   
1              20        2023  XDA18116-89-4A  consumables   
2               4        2023  ABE59463-05-7E     software   
3              17        2023  ZMM00836-65-0C  consumables   
4              21        2023  BZX55559-12-3X  consumables   
...           ...         ...             ...          ...   
54634          17        2023  UCA28532-31-0B     software   
54635           6        2023  EBD29277-23-7U  consumables   
54636           9        2023  EBD30093-55-9Z  consumables   
54637           3        2023  XMZ38239-49-3D    furniture   
54638           1        2023  ZCZ05589-50-5D     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
0                wall art     1096.80     762.71         7.50   105   
1                    pens       24.95      15.09         1.49    21   
2      project management       13.52       7.86         1.68    39   
3                    pens       36.42      24.85         1.23    29   
4                    misc      195.10     108.17        46.43    20   
...                   ...         ...        ...          ...   ...   
54634           logistics       83.13      51.60         2.25    33   
54635    kitchen supplies      206.59     175.46        11.70    47   
54636   bathroom supplies       65.66      57.31         4.16   475   
54637              stands        1.48       1.22        18.04   112   
54638  project management        3.01       1.61         2.07  1031   

       line_number  line_subtotal  
0                1      115164.00  
1                0         523.95  
2                6         527.28  
3                3        1056.18  
4                1        3902.00  
...            ...            ...  
54634            8        2743.29  
54635            0        9709.73  
54636            6       31188.50  
54637            0         165.76  
54638            5        3103.31  

[54639 rows x 19 columns]

In [16]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df["weight"] = df["unit_weight"] * df["qty"]
df["shipping_price"] = df["weight"].apply(lambda x: 7 * x if x > 50 else 10 * x)
df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ...     category         subcategory  \
0              17        2023  ...        decor            wall art   
1              20        2023  ...  consumables                pens   
2               4        2023  ...     software  project management   
3              17        2023  ...  consumables                pens   
4              21        2023  ...  consumables                misc   
...           ...         ...  ...          ...                 ...   
54634          17        2023  ...     software           logistics   
54635           6        2023  ...  consumables    kitchen supplies   
54636           9        2023  ...  consumables   bathroom supplies   
54637           3        2023  ...    furniture              stands   
54638           1        2023  ...     software  project management   

      unit_price  unit_cost  unit_weight   qty  line_number  line_subtotal  \
0        1096.80     762.71         7.50   105            1      115164.00   
1          24.95      15.09         1.49    21            0         523.95   
2          13.52       7.86         1.68    39            6         527.28   
3          36.42      24.85         1.23    29            3        1056.18   
4         195.10     108.17        46.43    20            1        3902.00   
...          ...        ...          ...   ...          ...            ...   
54634      83.13      51.60         2.25    33            8        2743.29   
54635     206.59     175.46        11.70    47            0        9709.73   
54636      65.66      57.31         4.16   475            6       31188.50   
54637       1.48       1.22        18.04   112            0         165.76   
54638       3.01       1.61         2.07  1031            5        3103.31   

        weight  shipping_price  
0       787.50         5512.50  
1        31.29          312.90  
2        65.52          458.64  
3        35.67          356.70  
4       928.60         6500.20  
...        ...             ...  
54634    74.25          519.75  
54635   549.90         3849.30  
54636  1976.00        13832.00  
54637  2020.48        14143.3

In [17]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df["total_price"] = (df["line_subtotal"] + df["shipping_price"]) * 1.0925
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number  line_subtotal  weight  shipping_price  \
0         7.50  105            1      115164.00  787.50         5512.50   
1         1.49   21            0         523.95   31.29          312.90   
2         1.68   39            6         527.28   65.52          458.64   
3         1.23   29            3        1056.18   35.67          356.70   
4        46.43   20            1        3902.00  928.60         6500.20   

     total_price  
0  131839.076250  
1     914.258625  
2    1077.117600  
3    1543.571400  
4   11364.403500  

[5 rows x 22 columns]

In [18]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df["total_cost"] = df["unit_cost"] *  df["qty"] + df["shipping_price"]
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number  line_subtotal  weight  shipping_price    total_price  \
0            1      115164.00  787.50         5512.50  131839.076250   
1            0         523.95   31.29          312.90     914.258625   
2            6         527.28   65.52          458.64    1077.117600   
3            3        1056.18   35.67          356.70    1543.571400   
4            1        3902.00  928.60         6500.20   11364.403500   

   total_cost  
0    85597.05  
1      629.79  
2      765.18  
3     1077.35  
4     8663.60  

[5 rows x 23 columns]

In [19]:
# Create a column for the profit of each line using line cost and line price
df["profit"] = df["total_price"] - df["total_cost"]
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

   line_subtotal  weight  shipping_price    total_price  total_cost  \
0      115164.00  787.50         5512.50  131839.076250    85597.05   
1         523.95   31.29          312.90     914.258625      629.79   
2         527.28   65.52          458.64    1077.117600      765.18   
3        1056.18   35.67          356.70    1543.571400     1077.35   
4        3902.00  928.60         6500.20   11364.403500     8663.60   

         profit  
0  46242.026250  
1    284.468625  
2    311.937600  
3    466.221400  
4   2700.803500  

[5 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [20]:
# Check your work using the totals above
df.shape


(54639, 24)

In [21]:
unique_ids = df["order_id"].nunique()
unique_ids

10000

In [22]:
order_2742071 = df[df["order_id"] == 2742071]["total_price"].sum()
order_2742071

np.float64(152811.897075)

In [23]:
order_2173913 = df[df["order_id"] == 2173913]["total_price"].sum()
order_2173913

np.float64(162388.7193)

In [24]:
order_6128929 = df[df["order_id"] == 6128929]["total_price"].sum()
order_6128929

np.float64(923441.24045)

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [29]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
top_5_spending = df[df['client_id'].isin(top_5_list)].groupby('client_id')['total_price'].sum()
top_5_spending


client_id
24741    8.226889e+07
33615    8.377309e+06
38378    1.290655e+07
46820    9.743794e+06
66037    1.025951e+07
Name: total_price, dtype: float64

In [30]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit.
#Filter for top five IDs.
filtered_df = df[df['client_id'].isin(top_5_list)]

# Summarize data for the top clients
summary_df = filtered_df.groupby('client_id')[['qty', 'shipping_price', 'total_price', 'profit']].sum()
summary_df[["total_price", "profit"]] = summary_df[["total_price", "profit"]] / 1000000
summary_df


qty  shipping_price  total_price     profit
client_id                                                
24741      239862      5126448.37    82.268892  36.579992
33615       64313      1828984.89     8.377309   2.201995
38378       73667      3429455.40    12.906551   3.271830
46820       75768      1601448.84     9.743794   2.736603
66037       43018      1395151.85    10.259515   3.255032

In [31]:
# Format the data and rename the columns to names suitable for presentation.
summary_df.columns = ['total_units', 'total_shipping', 'total_revenue', 'total_profit']

# Define the money columns.
summary_df = summary_df.rename(columns={
    "total_shipping": "total_shipping_USD",
    "total_revenue": "total_revenue_USD",
    "total_profit": "total_profit_USD"
})
summary_df

# Define a function that converts a dollar amount to millions.
## DID THIS ABOVE INSTEAD

# Apply the currency_format_millions function to only the money columns. 
## DID THIS ABOVE

# Rename the columns to reflect the change in the money format. 
## DID THIS ABOVE



total_units  total_shipping_USD  total_revenue_USD  \
client_id                                                       
24741           239862          5126448.37          82.268892   
33615            64313          1828984.89           8.377309   
38378            73667          3429455.40          12.906551   
46820            75768          1601448.84           9.743794   
66037            43018          1395151.85          10.259515   

           total_profit_USD  
client_id                    
24741             36.579992  
33615              2.201995  
38378              3.271830  
46820              2.736603  
66037              3.255032

In [32]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
profit_sorted_df = summary_df.sort_values(["total_profit_USD"], ascending=False)
profit_sorted_df

total_units  total_shipping_USD  total_revenue_USD  \
client_id                                                       
24741           239862          5126448.37          82.268892   
38378            73667          3429455.40          12.906551   
66037            43018          1395151.85          10.259515   
46820            75768          1601448.84           9.743794   
33615            64313          1828984.89           8.377309   

           total_profit_USD  
client_id                    
24741             36.579992  
38378              3.271830  
66037              3.255032  
46820              2.736603  
33615              2.201995